File: server/scraper/websites/rtvSlo.ts
import { INews } from '../../model/News';
import * as Db from '../../db/db';
import puppeteer from 'puppeteer';
import { parse } from 'date-fns';
//TODO :: scrape sub title and quot(figure)
async function rtvSlo(n: number) {
  const newsList: INews[] = [];

  const browser = await puppeteer.launch({ headless: 'new' });
  const page = await browser.newPage();

  await page.goto('https://www.rtvslo.si/novice');

  const links = await page.$$('div.md-news');

  for (let i = 0; i < links.length && newsList.length < n; i++) {
    const link = links[i];

    const url = await link.$eval('a', (anchor) =>
      anchor ? `https://www.rtvslo.si${anchor.getAttribute('href')}` : null
    );

    if (url) {
      const articlePage = await browser.newPage();
      await articlePage.goto(url);

      const title = await articlePage.$eval(
        'header.article-header h1',
        (element) => element?.textContent?.trim() || ''
      );

      const dateString = await articlePage.$eval(
        'meta[name="published_date"]',
        (el) => el?.getAttribute('content') || ''
      );
      const date = parse(dateString, "yyyy-MM-dd'T'HH:mm:ss'Z'", new Date());

      const contentTmp = await articlePage.$eval(
        'article.article',
        (element) => element.textContent?.trim() || ''
      );
      const content = contentTmp.replace(/[\t\n]/g, '').trim();

      const metaElements = await articlePage.$$('meta.elastic[name="author"]');
      const authors = await Promise.all(
        metaElements.map(async (el) => {
          const content = await el.getProperty('content');
          const author = await content.jsonValue();
          return author ?? '';
        })
      );

      const locationTmp = await articlePage.$eval(
        'div.place-source',
        (el) => (el as HTMLElement).textContent?.trim() || ''
      );

      const coords: [number, number] =
        locationTmp !== 'MMC RTV SLO' && locationTmp !== ''
          ? await Db.Util.toCoords(locationTmp.split(', ')[0])
          : [0, 0];

      const categoriesString = await articlePage.$eval(
        'meta[name="keywords"]',
        (el) => el?.getAttribute('content')?.toLowerCase() || ''
      );
      const categories = categoriesString ? categoriesString.split(',') : [];

      const news: INews = {
        title,
        url,
        date,
        authors,
        content,
        categories,
        views: [],
        location: {
          type: 'Point',
          coordinates: coords,
        },
      };
      newsList.push(news);

      await articlePage.close();
    }
  }

  await browser.close();

  return newsList;
}

export = rtvSlo;

File: desktop-app/src/jvmMain/kotlin/scraper/rtvSlo.kt
import org.example.model.INews
import org.example.model.Location
import org.openqa.selenium.By
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.chrome.ChromeOptions
import java.text.SimpleDateFormat
import java.util.Date

fun getRtvSlo(numArticlesToOpen: Int): List<INews> {

    val options = ChromeOptions()
    options.addArguments("--headless=new");
    val driver: WebDriver = ChromeDriver(options)

    driver.get("https://www.rtvslo.si/novice")

    val articles: List<WebElement> = driver.findElements(By.className("md-news"))

    // Create an empty list to store the news articles
    val newsList = mutableListOf<INews>()

    // Loop through the specified number of news articles and extract the data
    for (i in 0 until numArticlesToOpen) {
        val article: WebElement = articles[i]

        val url: String = article.findElement(By.cssSelector("a")).getAttribute("href")
        driver.get(url)

        var metaTag: WebElement = driver.findElement(By.cssSelector("meta[name='title']"))
        val title: String = metaTag.getAttribute("content")

        metaTag = driver.findElement(By.cssSelector("meta[name='published_date']"))
        val dateString: String = metaTag.getAttribute("content")
        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'")
        val date: Date = dateFormat.parse(dateString)

        val content: String = driver.findElement(By.cssSelector("article.article")).text

        val metaElements = driver.findElements(By.cssSelector("meta.elastic[name='author']"))
        val authors = metaElements.map { it.getAttribute("content") }

        val element = driver.findElement(By.cssSelector("div.place-source"))
        val location = element.text.trim().split(" - ")[0]

        metaTag = driver.findElement(By.cssSelector("meta.elastic[name='keywords']"))
        val categoriesString: String = metaTag.getAttribute("content")
        val categories = categoriesString.split(",")

        // Create an INews object and add it to the newsList
        val news = INews(title, url, date, authors, content, categories, Location(
            type = "Point",
            coordinates = Pair(0.0,0.0),
        )
        )
        newsList.add(news)

        driver.navigate().back()
    }

    driver.quit()

    return newsList
}
